controllers/share.go

```diff
@@ -0,0 +1,185 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws/arn"
+	"github.com/go-logr/logr"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	capi "sigs.k8s.io/cluster-api/api/v1beta1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/giantswarm/k8smetadata/pkg/annotation"
+
+	"github.com/giantswarm/aws-network-topology-operator/pkg/aws"
+	"github.com/giantswarm/aws-network-topology-operator/pkg/util/annotations"
+)
+
+const FinalizerResourceShare = "network-topology.finalizers.giantswarm.io/share"
+
+//counterfeiter:generate . RAMClient
+type RAMClient interface {
+	ApplyResourceShare(context.Context, aws.ResourceShare) error
+	DeleteResourceShare(context.Context, string) error
+}
+
+type ShareReconciler struct {
+	ramClient     RAMClient
+	clusterClient ClusterClient
+}
+
+func NewShareReconciler(clusterClient ClusterClient, ramClient RAMClient) *ShareReconciler {
+	return &ShareReconciler{
+		ramClient:     ramClient,
+		clusterClient: clusterClient,
+	}
+}
+
+func (r *ShareReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&capi.Cluster{}).
+		Complete(r)
+}
+
+func (r *ShareReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := r.getLogger(ctx)
+
+	logger.Info("Reconciling")
+	defer logger.Info("Done reconciling")
+
+	cluster, err := r.clusterClient.Get(ctx, req.NamespacedName)
+	if k8serrors.IsNotFound(err) {
+		logger.Info("cluster no longer exists")
+		return ctrl.Result{}, nil
+	}
+	if err != nil {
+		logger.Error(err, "failed to get cluster")
+		return ctrl.Result{}, err
+	}
+
+	if !cluster.DeletionTimestamp.IsZero() {
+		logger.Info("Reconciling delete")
+		return r.reconcileDelete(ctx, cluster)
+	}
+
+	return r.reconcileNormal(ctx, cluster)
+}
+
+func (r *ShareReconciler) reconcileDelete(ctx context.Context, cluster *capi.Cluster) (ctrl.Result, error) {
+	logger := r.getLogger(ctx)
+
+	err := r.ramClient.DeleteResourceShare(ctx, getResourceShareName(cluster))
+	if err != nil {
+		logger.Error(err, "failed to apply resource share")
+		return ctrl.Result{}, err
+	}
+
+	err = r.clusterClient.RemoveFinalizer(ctx, cluster, FinalizerResourceShare)
+	if err != nil {
+		logger.Error(err, "failed to remove finalizer")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ShareReconciler) reconcileNormal(ctx context.Context, cluster *capi.Cluster) (ctrl.Result, error) {
+	logger := r.getLogger(ctx)
+
+	if annotations.GetAnnotation(cluster, annotation.NetworkTopologyModeAnnotation) != annotation.NetworkTopologyModeGiantSwarmManaged {
```

Comment:
Wouldn't this apply to `reconcileDelete` as well? Like, why reconcile deletion when we skip on `reconcileNormal`? I'd move this outside of this function.

Reply:
As far as I see, Mario implemented a different logic for `reconcileDelete` which checks implicitly if share reconciler should delete.

Reply:
Oh, I was looking at wrong line. Yes, it should also apply this to delete. I will move it to `Reconcile` function.